consensus/ibft/signer/signer.go

```diff
@@ -0,0 +1,336 @@
+	if err != nil {
+		return nil, err
+	}
+
+	header.ExtraData = packCommittedSealIntoExtra(
+		header.ExtraData,
+		committedSeal,
+	)
+
+	return header, nil
+}
+
+func (s *SignerImpl) VerifyCommittedSeals(
+	header *types.Header,
+	validators validators.Validators,
+	quorumSize int,
+) error {
+	extra, err := s.GetIBFTExtra(header)
+	if err != nil {
+		return err
+	}
+
+	hash, err := s.CalculateHeaderHash(header)
+	if err != nil {
+		return err
+	}
+
+	rawMsg := wrapCommitHash(hash[:])
+
+	numSeals, err := s.keyManager.VerifyCommittedSeals(
+		extra.CommittedSeal,
+		rawMsg,
+		validators,
+	)
+	if err != nil {
+		return err
+	}
+
+	if numSeals < quorumSize {
+		return ErrNotEnoughCommittedSeals
+	}
+
+	return nil
+}
+
+func (s *SignerImpl) VerifyParentCommittedSeals(
+	parent, header *types.Header,
+	parentValidators validators.Validators,
+	quorum int,
+	mustExist bool,
+) error {
+	parentCommittedSeals, err := s.GetParentCommittedSeals(header)
+	if err != nil {
+		return err
+	}
+
+	if parentCommittedSeals == nil {
+		// Throw error for the proposed header
+		if mustExist {
+			return ErrEmptyParentCommittedSeals
+		}
+
+		// Don't throw if the flag is unset for backward compatibility
+		// (for the past headers)
+		return nil
+	}
+
+	rawMsg := wrapCommitHash(parent.Hash[:])
+
+	numSeals, err := s.keyManager.VerifyCommittedSeals(
+		parentCommittedSeals,
+		rawMsg,
+		parentValidators,
+	)
+	if err != nil {
+		return err
+	}
+
+	if numSeals < quorum {
+		return ErrNotEnoughCommittedSeals
+	}
+
+	return nil
+}
+
+func (s *SignerImpl) SignIBFTMessage(msg []byte) ([]byte, error) {
+	return s.keyManager.SignIBFTMessage(msg)
+}
+
+func (s *SignerImpl) initIbftExtra(
+	header *types.Header,
+	validators validators.Validators,
+	parentCommittedSeal Sealer,
+) {
+	putIbftExtra(header, &IstanbulExtra{
+		Validators:          validators,
+		ProposerSeal:        nil,
+		CommittedSeal:       s.keyManager.NewEmptyCommittedSeal(),
+		ParentCommittedSeal: parentCommittedSeal,
+	})
+}
+
+func (s *SignerImpl) CalculateHeaderHash(header *types.Header) (types.Hash, error) {
+	filteredHeader, err := s.filterHeaderForHash(header)
+	if err != nil {
+		return types.ZeroHash, err
+	}
+
+	return calculateHeaderHash(filteredHeader), nil
+}
+
+func (s *SignerImpl) filterHeaderForHash(header *types.Header) (*types.Header, error) {
+	clone := header.Copy()
+
+	extra, err := s.GetIBFTExtra(clone)
+	if err != nil {
+		return nil, err
+	}
+
+	// This will effectively remove the Seal and Committed Seal fields,
+	// while keeping proposer vanity and validator set
+	// because extra.Validators, extra.ParentCommittedSeal is what we got from `h` in the first place.
+	s.initIbftExtra(clone, extra.Validators, extra.ParentCommittedSeal)
+
+	return clone, nil
+}
+
+// extractParentCommittedSeals extracts Parent Committed Seals from IBFT Extra in Header
+func (s *SignerImpl) GetParentCommittedSeals(header *types.Header) (Sealer, error) {
```

Comment:
Comment and method mismatch

Reply:
Fixed at 9014bab9dff8fdb9e97c77830b0b054b6db3a23d